### PR TITLE
Log warnings in COPY if no files match or files can't be found

### PR DESCRIPTION
**Pull request by mxm**

### CODE REVIEW

File: sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java [hunk 1]
```diff
@@ -201,18 +201,27 @@ private void advanceToNextFileInput() throws IOException {
...
+            InputStream stream = fileInput.getStream(uri);
+            if (stream != null) {
+                currentReader = createBufferedReader(stream);
+                currentLineNumber = 0;
+                lineProcessor.readFirstLine(currentUri, inputFormat, currentReader);
+            } else {
+                LOGGER.warn("Couldn't open reader for URI '{}'", uri);
+            }
+        } catch (IOException e) {
+            LOGGER.warn("{} when reading from URI '{}': {}",
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java]: why not logging the complete exception? wouldn't that help in many cases?

File: sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java [hunk 2]
```diff
@@ -201,18 +201,27 @@ private void advanceToNextFileInput() throws IOException {
...
 
     private void initCurrentReader(FileInput fileInput, URI uri) throws IOException {
-        InputStream stream = fileInput.getStream(uri);
-        if (stream != null) {
-            currentReader = createBufferedReader(stream);
-            currentLineNumber = 0;
-            lineProcessor.readFirstLine(currentUri, inputFormat, currentReader);
+        try {
+            InputStream stream = fileInput.getStream(uri);
+            if (stream != null) {
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java]: afaik the stream cannot be null anymore with your change

File: sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java (hunk 1 above)
Comment (mxm) [reply sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java]: I wanted to keep this readable and avoid spaming the log with the full stack trace. I don't think the full exception really helps. The type and the message should be enough.

Comment (mxm) [reply sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java]: Maybe we could log the full exception with debug flag.

File: sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java (hunk 2 above)
Comment (seut) [reply sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java]: right, S3Input still returns null. can you please mark the `getStream` method as Nullable?

File: sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java (hunk 1 above)
Comment (seut) [reply sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java]: I fine with your solution, I wouldn't add extra complexity (branches) just for that.

Comment (mxm) [reply sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java]: OK, so I'll keep it as it is?

File: sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java [hunk 3]
```diff
@@ -201,18 +201,27 @@ private void advanceToNextFileInput() throws IOException {
         Predicate<URI> uriPredicate = generateUriPredicate(fileInput, fileUri.globPredicate);
         List<URI> uris = getUris(fileInput, fileUri.uri, fileUri.preGlobUri, uriPredicate);
 
-        if (uris.size() > 0) {
+        if (uris.isEmpty()) {
+            LOGGER.warn("No files matched the pattern '{}'", fileUri.uri);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java]: it's a common case that a file is only on a single node - I don't think that we should log a warning on all others.

Comment (mxm) [reply sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java]: If you use `/dir/*` then you won't have any missing files. If you use a hardcoded URL then I think you're doing something wrong. For example, you might have forgotten to copy the file to the node.

**Comment by mxm** [comment]
The consensus was to improve the logging regardless of the `COPY` changes. The result set which contains the errors will only be returned if you use `COPY ... RETURN SUMMARY`. If we make this the default, then I totally agree that logging is not necessary.

The main problem here seems to be that we don't have a separation for operational warnings/errors and user-related messages. 

I still would like to change the exception handling but I can remove the log messages because there seems to be no consent on the matter anymore.

File: sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java (hunk 3 above)
Comment (mxm) [reply sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java]: Have to correct myself, of course it also logs for `/dir/*` here. That is questionable.

**Comment by mxm** [comment]
Could we keep the logging of `IOExceptions` during reading or opening the file? I think that would make sense. These are real errors because these files should be read.

**Comment by mfussenegger** [comment]
> The consensus was to improve the logging regardless of the COPY changes. The result set which contains the errors will only be returned if you use COPY ... RETURN SUMMARY. If we make this the default, then I totally agree that logging is not necessary.

I don't think that we want to tell users to take a look at the logs if they've trouble with COPY, we want to tell them to use `RETURN SUMMARY`. 

Logs, especially warning or error, should be for administration or operations. There shouldn't be any "logs for users". As a user you don't want to start messing with logs if you've a 20+ nodes cluster.

**Comment by mxm** [comment]
The user log wouldn't have to be a log file. The `RETURN SUMMARY` will be a kind of user log.

Do you agree that file reading errors should be logged? There might be some issue with the file system that administrators should look into.

**Comment by mxm** [comment]
Closing because there is no need for logging at the moment with the `COPY ... RETURN SUMMARY` coming up.

**Pull request closed without merging**